underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java

```diff
@@ -0,0 +1,486 @@
+    }
+  }
+
+  /**
+   * Gets the parent key of the input key, or null if no parent exists
+   * @param key the key to get the parent of
+   * @return the the parent key
+   */
+  private String getParentKey(String key) {
+    int separatorIndex = key.lastIndexOf(PATH_SEPARATOR);
+    if (separatorIndex >= 0) {
+      return key.substring(0, separatorIndex);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Determines if the key represents a folder. If false is returned, it is not guaranteed that the
+   * path exists.
+   * @param key the key to check
+   * @return S3Object containing metadata
+   */
+  private boolean isFolder(String key) {
+    key = key.endsWith(PATH_SEPARATOR) ? key.substring(0, key.length() - 1) : key;
+    // Root is always a folder
+    if (isRoot(key)) {
+      return true;
+    }
+    try {
+      String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+      mClient.getObjectDetails(mBucketName, keyAsFolder);
+      // If no exception is thrown, the key exists as a folder
+      return true;
+    } catch (ServiceException se) {
+      return false;
+    }
+  }
+
+  /**
+   * Checks if the key is the root.
+   * @param key the key to check
+   * @return true if the key is the root, false otherwise
+   */
+  private boolean isRoot(String key) {
+    return key.equals(Constants.HEADER_S3N + mBucketName)
+        || key.equals(Constants.HEADER_S3N + mBucketName + PATH_SEPARATOR);
+  }
+
+  /**
+   * Lists the files in the given path, the paths will be their logical names and not contain the
+   * folder suffix
+   * @param path the key to list
+   * @param recursive if true will list children directories as well
+   * @return an array of the file and folder names in this directory
+   * @throws IOException
+   */
+  private String[] listInternal(String path, boolean recursive) throws IOException {
+    try {
+      path = stripPrefixIfPresent(path);
+      path = path.endsWith(PATH_SEPARATOR) ? path : path + PATH_SEPARATOR;
+      path = path.equals(PATH_SEPARATOR) ? "" : path;
+      // Gets all the objects under the path, because we have no idea if there are non Tachyon
+      // managed "directories"
+      S3Object[] objs = mClient.listObjects(mBucketName, path, "");
+      if (recursive) {
+        String[] ret = new String[objs.length];
+        for (int i = 0; i < objs.length; i ++) {
+          // Remove parent portion of the key
+          String child = objs[i].getKey().substring(path.length());
+          // Prune the special folder suffix
+          child =
+              child.endsWith(FOLDER_SUFFIX) ? child.substring(0,
+                  child.length() - FOLDER_SUFFIX.length()) : child;
+          ret[i] = child;
+        }
+        return ret;
+      }
+      // Non recursive list
+      Set<String> children = new HashSet<String>();
+      for (S3Object obj : objs) {
+        // Remove parent portion of the key
+        String child = obj.getKey().substring(path.length());
```

Comment:
This logic to derive child name could be a separate function in this class, and I highly recommend to unit test this function. Maybe not for this PR though.